### ISSUE TITLE
crash on hyperopt - amount too high on some pairs while optimizing, getting absurd numbers

### ISSUE THREAD

**Issue by div4r4k** [freqtrade/freqtrade#11441]
Hello, I can´t find similar posts with the problem I am experiencing.

## Describe your environment

  * Operating system: Linux Mint 22.1, kernel 6.8.0-40
  * Python Version: 3.12.3
  * CCXT version:  4.4.59
  * Freqtrade Version: freqtrade 2025.2-dev-840f074c6
  
## Describe the problem:

When I hyperopt some strategies with leverage, the output start to show oscene profit numbers, and in some time crashes with something like:

ERROR - Amount 4098091.1267917478 too high for XMR/USDT:USDT (sometime is another pair)

the command line is:
freqtrade hyperopt --config user_data/config.json --hyperopt-loss MultiMetricHyperOptLoss --strategy RsiquiV3_2 -e 500 --spaces {roi,stoploss,trailing,trades}  -j 24  --timerange 20250101-20250226  --enable-protections --freqaimodel LightGBMRegressor 

my config file:

```
{
    "$schema": "https://schema.freqtrade.io/schema.json",
    "max_open_trades": 5,
    "stake_currency": "USDT",
    "stake_amount": "unlimited",
    "tradable_balance_ratio": 0.5,
    "fiat_display_currency": "USD",
    "dry_run": true,
    "dry_run_wallet": 100,
    "cancel_open_orders_on_exit": false,
	"amend_last_stake_amount": true,
    "trading_mode": "futures",
    "margin_mode": "isolated",
	"liquidation_buffer": 0.05,
    "unfilledtimeout": {
        "entry": 10,
        "exit": 10,
        "exit_timeout_count": 0,
        "unit": "minutes"
    },
    "entry_pricing": {
        "price_side": "other",
        "use_order_book": true,
        "order_book_top": 1,
        "price_last_balance": 0.0,
        "check_depth_of_market": {
            "enabled": false,
            "bids_to_ask_delta": 1
        }
    },
    "exit_pricing":{
        "price_side": "other",
        "use_order_book": true,
        "order_book_top": 1
    },
    "exchange": {
        "name": "binance",
        "key": "",
        "secret": "",
        "password": "",
        "ccxt_config": {
            "enableRateLimit": true,
            "options": { "defaultType": "future"}
            },
        "ccxt_async_config": {
            "enableRateLimit": true,
            "rateLimit": 200,
            "options": { "defaultType": "future"}
            },
        "pair_whitelist": [
		"BTC/USDT:USDT",
        "BCH/USDT:USDT",
        "ETH/USDT:USDT",
        "LINK/USDT:USDT",
        "LTC/USDT:USDT",
        "SOL/USDT:USDT",
        "BNB/USDT:USDT",
        "XRP/USDT:USDT",
        "ADA/USDT:USDT",
        "DOT/USDT:USDT",
        "ETC/USDT:USDT",
        "ALGO/USDT:USDT",
		"XMR/USDT:USDT",
		"AVAX/USDT:USDT",
		"AAVE/USDT:USDT",
		"SUI/USDT:USDT",
		"TON/USDT:USDT",
		"NEAR/USDT:USDT",
        "RAY/USDT:USDT",
        "HBAR/USDT:USDT",
        "INJ/USDT:USDT",
        "ROSE/USDT:USDT",
        ],
        "pair_blacklist": [
        ]
    },
    "pairlists": [
        {
            "method": "StaticPairList",
            "number_assets": 20,
            "sort_key": "percentage",
            "min_value": 0,
            "refresh_period": 3600,
			"lookback_timeframe": "1h",
			"lookback_period": 72
        }
    ],
"freqai": {
        "enabled": true,
        "purge_old_models": 2,
        "train_period_days": 30,
        "backtest_period_days": 14,
        "identifier" : "unique-id",
        "feature_parameters" : {
            "include_timeframes": ["5m"],
            "include_corr_pairlist": [
                "BTC/USDT:USDT",
                "BCH/USDT:USDT",
                "ETH/USDT:USDT",
                "LINK/USDT:USDT",
                "LTC/USDT:USDT",
                "SOL/USDT:USDT",
                "BNB/USDT:USDT",
                "XRP/USDT:USDT",
                "ADA/USDT:USDT",
                "DOT/USDT:USDT",
                "ETC/USDT:USDT",
                "ALGO/USDT:USDT",
		        "XMR/USDT:USDT",
		        "AVAX/USDT:USDT",
		        "AAVE/USDT:USDT",
		        "SUI/USDT:USDT",
		        "TON/USDT:USDT",
		        "NEAR/USDT:USDT",
                "RAY/USDT:USDT",
                "HBAR/USDT:USDT",
                "INJ/USDT:USDT",
                "ROSE/USDT:USDT",
            ],
            "label_period_candles": 24,
            "include_shifted_candles": 2,
            "indicator_periods_candles": [10, 20]
        },
        "data_split_parameters" : {
            "test_size": 0.25
        },
        "model_training_parameters": {}
        },
"telegram": {
    "enabled": true,
    "token": "",
    "chat_id": "",
    "allow_custom_messages": true,
    "notification_settings": {
        "status": "silent",
        "warning": "on",
        "startup": "off",
        "entry": "silent",
        "entry_fill": "on",
        "entry_cancel": "silent",
        "exit": {
            "roi": "silent",
            "emergency_exit": "on",
            "force_exit": "on",
            "exit_signal": "silent",
            "trailing_stop_loss": "on",
            "stop_loss": "on",
            "stoploss_on_exchange": "on",
            "custom_exit": "silent",  // custom_exit without specifying an exit reason
            "partial_exit": "on",
            // "custom_exit_message": "silent",  // Disable individual custom exit reasons
            "*": "off"  // Disable all other exit reasons
        },
        // "exit": "off",  // Simplistic configuration to disable all exit messages
        "exit_cancel": "on",
        "exit_fill": "off",
        "protection_trigger": "off",
        "protection_trigger_global": "on",
        "strategy_msg": "off",
        "show_candle": "off"
    },
    "reload": true,
    "balance_dust_level": 0.01
},
    "api_server": {
        "enabled": true,
        "listen_ip_address": "127.0.0.1",
        "listen_port": 8084,
        "verbosity": "error",
        "enable_openapi": false,
        "jwt_secret_key": "",
        "ws_token": "",
        "CORS_origins": [],
        "username": "freqtrade",
        "password": "freqtrade"
    },
    "bot_name": "futures",
    "initial_state": "running",
    "force_entry_enable": false,
    "internals": {
        "process_throttle_secs": 5
    }
}
```

And here's the strategy:

``` python
import numpy
import talib.abstract as ta
import freqtrade.vendor.qtpylib.indicators as qtpylib
from pandas import DataFrame
from datetime import datetime
from typing import Optional, Tuple, Union
from freqtrade.persistence import Trade
from freqtrade.strategy.interface import IStrategy
from freqtrade.strategy import (IStrategy, DecimalParameter, IntParameter, CategoricalParameter, BooleanParameter)


class RsiquiV3_2(IStrategy):
    INTERFACE_VERSION = 3

    can_short = True
    timeframe = '5m'
    stoploss = -0.05
    trailing_stop = False
    max_open_trades = 10

    minimal_roi = {
      '0': 0.02
    }

    rsi_entry_long  = IntParameter(0,  50,  default=50, space='buy',  optimize=True)
    rsi_entry_short = IntParameter(50, 100, default=50, space='buy',  optimize=True)
    rsi_exit_long   = IntParameter(50, 100, default=95, space='sell', optimize=True)
    rsi_exit_short  = IntParameter(0,  50,  default=43, space='sell', optimize=True)

    @property
    def plot_config(self):
        plot_config = {}

        plot_config['main_plot'] = {
        }
        plot_config['subplots'] = {
            'Misc': {
                'rsi': {},
                'rsi_gra' : {},
            },
        }

        return plot_config

    def populate_indicators(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        dataframe['rsi'] = ta.RSI(dataframe, timeperiod=14)
        dataframe['rsi_gra'] = numpy.gradient(dataframe['rsi'], 60)

        return dataframe

    def populate_entry_trend(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        dataframe.loc[
            (
                (dataframe['rsi'] < self.rsi_entry_long.value) &
                qtpylib.crossed_above(dataframe['rsi_gra'], 0)
            ),
        'enter_long'] = 1

        dataframe.loc[
            (
                (dataframe['rsi'] > self.rsi_entry_short.value) &
                qtpylib.crossed_below(dataframe['rsi_gra'], 0)
            ),
        'enter_short'] = 1

        return dataframe

    def populate_exit_trend(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        dataframe.loc[
            (
                (dataframe['rsi'] > self.rsi_exit_long.value) &
                qtpylib.crossed_below(dataframe['rsi_gra'], 0)
            ),
        'exit_long'] = 1

        dataframe.loc[
            (
                (dataframe['rsi'] < self.rsi_exit_short.value) &
                qtpylib.crossed_above(dataframe['rsi_gra'], 0)
            ),
        'exit_short'] = 1

        return dataframe

    def adjust_trade_position(self, trade: Trade, current_time: datetime, current_rate: float, current_profit: float, min_stake: Optional[float], max_stake: float, current_entry_rate: float, current_exit_rate: float, current_entry_profit: float, current_exit_profit: float, **kwargs) -> Optional[float]:
        filled_entries = trade.select_filled_orders(trade.entry_side)
        count_of_entries = trade.nr_of_successful_entries
        
        if current_profit > 0.25 and trade.nr_of_successful_exits == 0:
            return -(trade.stake_amount / 4)
        if current_profit > 0.40 and trade.nr_of_successful_exits == 1:
            return -(trade.stake_amount / 3)
        
        if (current_profit > -0.15 and count_of_entries == 1) or \
        (current_profit > -0.3 and count_of_entries == 2) or \
        (current_profit > -0.6 and count_of_entries == 3):
            return None
        
        try:
            stake_amount = filled_entries[0].cost if filled_entries else None
            return stake_amount
        except Exception as exception:
            return None


    def leverage(self, pair: str, current_time: datetime, current_rate: float, proposed_leverage: float, max_leverage: float, side: str, **kwargs) -> float:
        window_size = 50
        base_leverage = 1

        dataframe, _ = self.dp.get_analyzed_dataframe(pair=pair, timeframe=self.timeframe)
        historical_close_prices = dataframe["close"].tail(window_size)
        historical_high_prices = dataframe["high"].tail(window_size)
        historical_low_prices = dataframe["low"].tail(window_size)

        rsi_values = ta.RSI(historical_close_prices, timeperiod=14)
        atr_values = ta.ATR(historical_high_prices, historical_low_prices, historical_close_prices, timeperiod=14)
        macd_line, signal_line, _ = ta.MACD(historical_close_prices, fastperiod=12, slowperiod=26, signalperiod=9)
        sma_values = ta.SMA(historical_close_prices, timeperiod=20)

        current_rsi = rsi_values[-1] if len(rsi_values) > 0 else 50.0
        current_atr = atr_values[-1] if len(atr_values) > 0 else 0.0
        current_macd = (macd_line[-1] - signal_line[-1]) if len(macd_line) > 0 and len(signal_line) > 0 else 0.0
        current_sma = sma_values[-1] if len(sma_values) > 0 else 0.0

        dynamic_rsi_low = numpy.nanmin(rsi_values) if len(rsi_values) > 0 and not numpy.isnan(numpy.nanmin(rsi_values)) else 30.0
        dynamic_rsi_high = numpy.nanmax(rsi_values) if len(rsi_values) > 0 and not numpy.isnan(numpy.nanmax(rsi_values)) else 70.0

        leverage_factors = {
            'long': {'increase': 10, 'decrease': 5},
            'short': {'increase': 10, 'decrease': 5},
        }

        if side == "long":
            base_leverage = base_leverage * leverage_factors['long']['increase'] if current_rsi < dynamic_rsi_low else base_leverage
            base_leverage = base_leverage * leverage_factors['long']['decrease'] if current_rsi > dynamic_rsi_high else base_leverage
            base_leverage = base_leverage * leverage_factors['long']['increase'] if current_macd > 0 else base_leverage
            base_leverage = base_leverage * leverage_factors['long']['decrease'] if current_rate < current_sma else base_leverage
        elif side == "short":
            base_leverage = base_leverage * leverage_factors['short']['increase'] if current_rsi > dynamic_rsi_high else base_leverage 
            base_leverage = base_leverage * leverage_factors['short']['decrease'] if current_rsi < dynamic_rsi_low else base_leverage
            base_leverage = base_leverage * leverage_factors['short']['increase'] if current_macd < 0 else base_leverage
            base_leverage = base_leverage * leverage_factors['short']['decrease'] if current_rate > current_sma else base_leverage

        adjusted_leverage = max(min(base_leverage, max_leverage), 1.0)

        return adjusted_leverage
```

I´d like to add that I am a complete newbie into python/coding, and can´t figure out by myself what I am doing wrong.

Thanks in advance.

**Comment by Vbranskil** [comment]
Hello @div4r4k 
Kindly send a direct message to the support helpdesk **@zendexk** on telegram to assist with this issue.

Thanks

**Comment by stash86** [comment]
use fixed stake to avoid such issue and to have realistic backtest result

**Comment by div4r4k** [comment]
@stash86 Thank you for the quick reply.
now using:

"stake_amount": 10,
 "tradable_balance_ratio": 0.99,


Backtested now and... well, still unrealistic to me (4k in less of a month with 100 usd for start), but the obscene numbers are gone :P hehe


Thanks again.

**Comment by xmatthias** [comment]
well i don't think using `np.gradient()` is safe and without lookahead bias.

doing a very simple test - taken from the documentation:
```
>>> import numpy as np
>>> f = np.array([1, 2, 4, 7, 11, 16])
>>> np.gradient(f)
array([1. , 1.5, 2.5, 3.5, 4.5, 5. ])
>>> f1 = np.array([1, 2, 4, 7, 11])
>>> np.gradient(f1)
array([1. , 1.5, 2.5, 3.5, 4. ])
```

The fact that the one number changes from 4.5 to 4 suggests that it's using the full timerange - so your insane numbers come from backtesting knowing what'll happen in the future.

---

I'm also not sure what you're actually trying to do
your config suggests the use of freqAI - but your strategy doesn't actually use it (in which case, you should disable it).
This is an in between state (configured but not used) i'm really not sure will be handled correctly - but is definitely not a supported way to run a non-freqAI bot.

---

`ERROR - Amount 4098091.1267917478 too high for XMR/USDT:USDT (sometime is another pair)`

This is not a "failing" error - but simply the result of too high stakes - resulting in the pair exceeding the maximum allowed position size on the exchange.
The bot will continue - just the trade in question can't be opened - as the requested stake would become too high.

**Comment by y-bowen** [comment]
You can use custom_stake_amount function：
```
max_stake = 100000
def custom_stake_amount(self, pair: str, current_time: datetime,
                            current_rate: float, proposed_stake: float,
                            min_stake: float, max_stake: float, side: str,
                            **kwargs) -> float:
        trade_amount = self.wallets.get_total_stake_amount() / \
            self.config["max_open_trades"]
        stake_amount = min(trade_amount, self.max_stake)
        return stake_amount
```

**Comment by xmatthias** [comment]
![Image](https://github.com/user-attachments/assets/eff135c8-42cf-46e9-8aa4-4caa2d35d5ef)

As mentioned above - the stake-amount falls "out of" the actual trading rules - so such an order would not be allowed to be placed on exchange - so the fact that freqtrade rejects this order is correct.

the "insane" profits are however caused by the lookahead bias of `np.gradient()` - which looks ahead by at least one candle - probably more (though i didn't test this).

**Comment by Dolfino** [comment]
2025-03-09 11:03:21,907 - freqtrade.plugins.pairlist.IPairList - INFO - Ignoring XMR/USDT from whitelist. Market is not active.
